File: java/RemoveNthNodeFromEndOfList.java
public class RemoveNthNodeFromEndOfList {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode temp = new ListNode(0);
        temp.next = head;
        head = temp;
        ListNode start = head;
        ListNode end = head;
        for (int i = 0; i < n; ++i, end = end.next) {
            if (end.next == null) {
                return head.next;
            }
        }

        for ( ; end.next != null; start = start.next, end = end.next) {
            ;
        }
        start.next = start.next.next;
        return head.next;
    }
}

File: cc/remove_nth_node_from_end_of_list.cc
#include "base_include.h"
#include "list_node.h"

class RemoveNthNodeFromEndOfList {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        ListNode *token = new ListNode(0);
        token->next = head;
        head = token;
        ListNode *slow = head;
        ListNode *fast = head;
        for (int i = 0; fast != NULL && i < n; ++i, fast = fast->next) {}
        if (NULL == fast) {
            return head;
        }
        for ( ; fast->next != NULL; slow = slow->next, fast = fast->next) {}
        token = slow->next;
        slow->next = slow->next->next;
        delete token;
        token = head;
        head = head->next;
        delete token;
        return head;
    }
};
